tensorflow/core/kernels/stage_op.cc

```diff
@@ -30,26 +30,82 @@ namespace {
 
 class Buffer : public ResourceBase {
  public:
-  explicit Buffer() {}
+    typedef std::vector<Tensor> Tuple;
 
-  typedef std::vector<Tensor> Tuple;
+ public:
+  explicit Buffer(int capacity) {
+    capacity_ = capacity;
+  }
+
+  bool HasBoundedCapacity() {
+    return capacity_ > 0;
+  }
+
+  bool IsFull() {
+    return buf_.size() >= capacity_;
+  }
 
   // the Buffer takes ownership of the Tuple
   void Put(Tuple* tuple) {
     mutex_lock l(mu_);
+
+    // If buffer capacity is bounded wait until elements have been removed
+    if(HasBoundedCapacity()) {
+      full_cond_var_.wait(l, [this]() {
+        return !this->IsFull();
+      });
+    }
+
     buf_.push_back(std::move(*tuple));
-    non_empty_cond_var_.notify_one();  // maybe possible to optimize by reducing
-                                       // how often this signal is sent
+
+    l.unlock();
+    // maybe possible to optimize by reducing
+    // how often this signal is sent
+    non_empty_cond_var_.notify_one();
   }
 
+  // Get tuple at front of the buffer
   void Get(Tuple* tuple) {  // TODO(zhifengc): Support cancellation.
     mutex_lock l(mu_);
-    while (buf_.empty()) {
-      non_empty_cond_var_.wait(l);
-    }
 
+    // Wait for data if the buffer is emptys
+    non_empty_cond_var_.wait(l, [this]() {
+      return !buf_.empty();
+    });
+
+    // Move data into the output tuple
     *tuple = std::move(buf_.front());
     buf_.pop_front();
+
+    notify_inserters_if_bounded(l);
+  }
+
+  // Return tuple at index
+  Status Peek(int index, Tuple* tuple) {
+    mutex_lock l(mu_);
+
+    // Wait if the requested index is not available
+    non_empty_cond_var_.wait(l, [&, this]() {
```

Comment:
It needs to capture both `index ` and `this`. I've changed the capture of `index` to capture by value as its just an int.

https://github.com/tensorflow/tensorflow/pull/9686/commits/a90c2e59a37acd2dd54d838261831aab34de3474